### DISCUSSION TITLE
Trigger event or new job on failure

### DISCUSSION THREAD

**Discussion by dmurphy-ambra** [jobrunr/jobrunr#329] [Q&A]
If a job fails, of course there is the RetryFilter to try it again.   However, if the job fails permanently (e.g. retries are exhausted), our system needs to send a notification to other systems to tell them that the job failed.  Is there a fluent way to trigger an event or new job "on failure"?

**Comment by rdehuyss** [comment]
There is not something build in out-of-the-box. You could create a filter that listens for the job state changes and if it failed (so when all of the retries are exhausted), perform some custom code. Would that help?

**Reply by dmurphy-ambra** [reply]
Sounds like it could work, I'll need to try and write such a filter.

**Reply by dmurphy-ambra** [reply]
@rdehuyss I'm doing the following to trigger an event after the number of retries expires, but this is the code for non-IoC.  Can you advise how I can alter this code to get my JobRequestHandler bean from the JobActivator?

The specific line I need to change is `Object jobRequestHandler = ReflectionUtils.newInstance(JobUtils.getJobClass(job.getJobDetails()));`

```
    public void onStateElection(Job job, JobState newState) {
        if (maxAmountOfRetriesReached(job) && isFailed(newState)) {
            logger.info("Job failed all retries {}", job);

            Object[] jobParameterValues = job.getJobDetails().getJobParameterValues();
            Object jobRequestHandler = ReflectionUtils.newInstance(JobUtils.getJobClass(job.getJobDetails()));
            getOnJobErrorMethod(jobRequestHandler).ifPresent(method -> {
                try {
                    method.invoke(jobRequestHandler, jobParameterValues);
                } catch (IllegalAccessException | InvocationTargetException e) {
                    logger.error("Cannot execute onJobError method for job {}", job);
                }
            });

        } else if (this.isFailed(newState)
                && !this.isProblematicExceptionAndMustNotRetry(newState)
                && !this.maxAmountOfRetriesReached(job)) {
            job.scheduleAt(
                Instant.now().plusSeconds(
                        this.retryPolicy.getSecondsToAdd(this.getFailureCount(job))
                ),
                String.format("Retry %d of %d", this.getFailureCount(job), this.getMaxNumberOfRetries(job))
            );
        }
    }
```

**Reply by dmurphy-ambra** [reply]
I managed to get the managed JobRequestHandler bean by injecting the JobActivator into my custom filter class during JobrunrConfiguration.  This isn't the cleanest as it means I'm injecting the Spring applicationContext, (generally not ideal) but it works. 

```
    @Bean
    public JobRunrConfiguration.JobRunrConfigurationResult initJobRunr(ApplicationContext applicationContext) {
        return JobRunrPro.configure()
                .useJobActivator(applicationContext::getBean)
                .useStorageProvider(jobRunrStorageProvider())
                .withJobFilter(new OnJobRequestFailureFilter(applicationContext::getBean))
                .useBackgroundJobServer()
                .useJmxExtensions()
                .useQueues(queueStandardPriority, queueHighPriority, queueStandardPriority, queueLowPriority)
                .useDashboard()
                .initialize();
    }
```

```
    public void onStateElection(Job job, JobState newState) {
        if (maxAmountOfRetriesReached(job) && isFailed(newState)) {
            logger.info("Job failed all retries {}", job);

            Object[] jobParameterValues = job.getJobDetails().getJobParameterValues();
            Object jobRequestHandler = jobActivator.activateJob(JobUtils.getJobClass(job.getJobDetails()));
            getOnJobErrorMethod(jobRequestHandler).ifPresent(method -> {
                try {
                    method.invoke(jobRequestHandler, jobParameterValues);
                } catch (IllegalAccessException | InvocationTargetException e) {
                    logger.error("Cannot execute onJobError method for job {}", job);
                }
            });

        }
```

**Reply by rdehuyss** [reply]
Since the JobFilters are singletons and only created once, I would not worry too much. But, I also may have some better idea's 😄

**Reply by dbouclier** [reply]
I get some failed jobs with exception like (JobNotFoundException) but I cannot put any specific logging or get notified in my monitoring when it's occurs, it's take time to see the issue in my code (for example, I forget to add a default constructor in the lambda argument) 

I know JobFilters are only available in pro version but can we just have a JobFailureHandler or something we can use in that case?

**Comment by tiarebalbi** [comment]
Any other thoughts for this discussion here? I still believe the support of publishers and subscribers would give any developer more flexibility in their implementation. Reference: #757
